pkg/hive/cell/metric.go

```diff
@@ -0,0 +1,151 @@
+	withMeta  pkgmetric.WithMetadata
+	collector prometheus.Collector
+)
+
+// Metric constructs a new metric cell.
+//
+// This cell type provides `S` to the hive as returned by `ctor`, it also makes each individual field
+// value available via the `hive-metrics` value group. Infrastructure components such as a registry,
+// inspection tool, or documentation generator can collect all metrics in the hive via this value group.
+//
+// The `ctor` constructor must return a struct or pointer to a struct of type `S`. The returned struct
+// must only contain public fields. All field types should implement the
+// `github.com/cilium/cilium/pkg/metrics/metric.WithMetadata`
+// and `github.com/prometheus/client_golang/prometheus.Collector` interfaces.
+func Metric[S any](ctor func() S) Cell {
+	caller := "caller"
+	if _, file, no, ok := runtime.Caller(1); ok {
+		caller = fmt.Sprintf("%s:%d", file, no)
+	}
+
+	var nilOut S
+	outTyp := reflect.TypeOf(nilOut)
+	if outTyp.Kind() == reflect.Ptr {
+		outTyp = outTyp.Elem()
+	}
+
+	if outTyp.Kind() != reflect.Struct {
+		panic(fmt.Sprintf(
+			"cell.Metric must be invoked with a constructor function that returns a struct or pointer to a struct, "+
+				"%s supplied a constructor which returns a %s",
+			caller,
+			outTyp.Kind(),
+		))
+	}
+
+	// Lets be strict for now, could lift this in the future if we ever need to
+	if outTyp.NumField() == 0 {
+		panic(fmt.Sprintf(
+			"cell.Metric must be invoked with a constructor function that returns exactly a struct with at least 1 "+
+				"metric, %s supplied a constructor which returns a struct with zero fields",
+			caller,
+		))
+	}
+
+	withMetaTyp := reflect.TypeOf(&withMeta).Elem()
+	collectorTyp := reflect.TypeOf(&collector).Elem()
+	for i := 0; i < outTyp.NumField(); i++ {
+		field := outTyp.Field(i)
+		if !field.IsExported() {
+			panic(fmt.Sprintf(
+				"The struct returned by the constructor passed to cell.Metric by %s has a private field '%s', which "+
+					"is not allowed. All fields on the returning struct must be exported",
+				caller,
+				field.Name,
+			))
+		}
+
+		if !field.Type.Implements(withMetaTyp) {
+			panic(fmt.Sprintf(
+				"The struct returned by the constructor passed to cell.Metric by %s has a field '%s', which is not "+
+					"metric.WithMetadata.",
+				caller,
+				field.Name,
+			))
+		}
+
+		if !field.Type.Implements(collectorTyp) {
+			panic(fmt.Sprintf(
+				"The struct returned by the constructor passed to cell.Metric by %s has a field '%s', which is not "+
+					"prometheus.Collector.",
+				caller,
+				field.Name,
+			))
+		}
+	}
+
+	return &metric[S]{
+		ctor: ctor,
+	}
+}
+
+type metric[S any] struct {
+	ctor func() S
+}
+
+type metricOut struct {
+	dig.Out
+
+	Metrics []pkgmetric.WithMetadata `group:"hive-metrics,flatten"`
+}
+
+func (m *metric[S]) provideMetrics(metricSet S) metricOut {
+	var metrics []pkgmetric.WithMetadata
+
+	value := reflect.ValueOf(metricSet)
+	typ := value.Type()
+	if typ.Kind() == reflect.Pointer {
```

Comment:
You can use https://go.dev/pkg/reflect/?m=old#Indirect to avoid this

Reply:
I tries, but that only works on a `reflect.Value` not on a `reflect.Type`.